src/plugins/fcrypt/fcrypt.c

```diff
@@ -0,0 +1,388 @@
+/**
+ * @file
+ *
+ * @brief filter plugin providing cryptographic operations
+ *
+ * @copyright BSD License (see doc/COPYING or http://www.libelektra.org)
+ *
+ */
+
+#ifndef HAVE_KDBCONFIG
+#include "kdbconfig.h"
+#endif
+#include "fcrypt.h"
+#include <gpg.h>
+#include <kdb.h>
+#include <kdberrors.h>
+#include <kdbtypes.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+enum fcryptGetState
+{
+	PREGETSTORAGE = 0,
+	POSTGETSTORAGE = 1
+};
+
+struct _fcryptState
+{
+	enum fcryptGetState getState;
+};
+typedef struct _fcryptState fcryptState;
+
+/**
+ * @brief Allocates a new string holding the name of the temporary file.
+ * @param file holds the path to the original file
+ * @returns an allocated string holding the name of the encrypted file. Must be freed by the caller.
+ */
+static char * getTemporaryFileName (const char * file)
+{
+	const size_t newFileAllocated = strlen (file) + 7;
+	char * newFile = elektraMalloc (newFileAllocated);
+	if (!newFile) return NULL;
+	snprintf (newFile, newFileAllocated, "%sXXXXXX", file);
+	mkstemp (newFile);
+	return newFile;
+}
+
+/**
+ * @brief lookup if the test mode for unit testing is enabled.
+ * @param conf KeySet holding the plugin configuration.
+ * @retval 0 test mode is not enabled
+ * @retval 1 test mode is enabled
+ */
+static int inTestMode (KeySet * conf)
+{
+	Key * k = ksLookupByName (conf, ELEKTRA_CRYPTO_PARAM_GPG_UNIT_TEST, 0);
+	if (k && !strcmp (keyString (k), "1"))
+	{
+		return 1;
+	}
+	return 0;
+}
+
+/**
+ * @brief Read number of total GPG recipient keys from the plugin configuration.
+ * @param config holds the plugin configuration
+ * @returns the number of GPG recipient keys.
+ */
+static size_t getRecipientCount (KeySet * config)
+{
+	Key * k;
+	size_t recipientCount = 0;
+	Key * root = ksLookupByName (config, ELEKTRA_CRYPTO_PARAM_GPG_KEY, 0);
+
+	if (!root) return 0;
+
+	// toplevel
+	if (strlen (keyString (root)) > 0)
+	{
+		recipientCount++;
+	}
+
+	ksRewind (config);
+	while ((k = ksNext (config)) != 0)
+	{
+		if (keyIsBelow (k, root))
+		{
+			recipientCount++;
+		}
+	}
+	return recipientCount;
+}
+
+/**
+ * @brief encrypt the file speicied at parentKey
+ * @param pluginConfig holds the plugin configuration
+ * @pararm parentKey holds the path to the file to be encrpted. Will hold an error description in case of failure.
+ * @retval 1 on success
+ * @retval -1 on error, errorKey holds an error description
+ */
+static int encrypt (KeySet * pluginConfig, Key * parentKey)
+{
+	const size_t recipientCount = getRecipientCount (pluginConfig);
+	if (recipientCount == 0)
+	{
+		ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_CRYPTO_CONFIG_FAULT, parentKey,
+				    "Missing GPG key (specified as %s) in plugin configuration.", ELEKTRA_CRYPTO_PARAM_GPG_KEY);
+		return -1;
+	}
+
+	const size_t testMode = inTestMode (pluginConfig);
+
+	char * tmpFile = getTemporaryFileName (keyString (parentKey));
```

Comment:
You should avoid to generate temporary files in this way, use mkstemp or similar instead.